### PR TITLE
Create basic Notification class

### PR DESCRIPTION
**Pull request by LilyFirefly**
We want a `Notification` class that we can register handlers with. These handlers will control notifying a user by email, text message, carrier pigeon or anything else required.

The `Notification` class will also need a way to define information that should be passed through to the handlers. For example, templates for rendering a message.

### CODE REVIEW

**Comment by LilyFirefly** [comment]
@meshy @KevinEtienne Does this look like what we want? Are there any features I've missed?

**Comment by meshy** [comment]
Should the handlers not be defining how they construct the message (template etc)?

**Comment by LilyFirefly** [comment]
That sounds like we might end up with a lot of handlers for a particular transport (eg email).

**Comment by LilyFirefly** [comment]
I was thinking that an instance of `Notification` would be the bit that is context-specific leaving the handlers to be fully generic.

**Comment by meshy** [comment]
That could work

**Comment by LilyFirefly** [comment]
@meshy @KevinEtienne Does this look like a good approach?

File: pigeon/notification.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    Notification holds the information necessary to notify a user.
+
+    Notifying a user requires at least one notification handler to be
+    registered using `register_handler`.
+    """
+    handlers = {}
+
+    def __init__(self, about):
+        """Store the object the notification is about."""
+        self.about = about
```
Comment (LilyFirefly) [inline pigeon/notification.py]: I'm not entirely happy with `about` as a name, but I'm not sure what else to call it.

Comment (meshy) [inline pigeon/notification.py]: I don't think that it's fair to prescribe that the notification will be about anything in particular.

For example, what would a "Welcome" email be "about"?

File: pigeon/notification.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    @classmethod
+    def register_handler(cls, **kwargs):
+        """Register a notification handler."""
+        cls.handlers.update(kwargs)
+
+    def notify(self, user):
+        """Use all registered handlers to notify a user."""
+        for handler in self.handlers.values():
+            handler(self.about, user)
```
Comment (meshy) [inline pigeon/notification.py]: possibly best to just pass through `self`

File: pigeon/notification.py (hunk 1 above)
Comment (kevinetienne) [inline pigeon/notification.py]: Maybe this could be more generic and be a `**context`?

File: pigeon/notification.py (hunk 2 above)
Comment (LilyFirefly) [inline pigeon/notification.py]: I'm not convinced - what other information could the `Notification` instance hold that the `handler` would care about?

Comment (meshy) [inline pigeon/notification.py]:
``` python
class WelcomeNotification(Notification):
    email_html_template = 'blah.html'
    email_template = 'blah.txt'
    sms_message = 'WELCOMZ! I R TEH WINNR!'
    pigeon_name = 'Murphy'
```

**Comment by kevinetienne** [comment]
Looks good to me, are you planning to add something else?

File: setup.cfg [hunk 3]
```diff
@@ -0,0 +1,5 @@
+[flake8]
+max-line-length = 90
+statistics = true
+application-import-names = pigeon
+import-order-style = google
```
Comment (meshy) [inline setup.cfg:5]:
```
[report]
show_missing = true
```

Comment (meshy) [inline setup.cfg:5]: This means that the obscure `-m` can go from the command in the makefile

**Comment by LilyFirefly** [comment]
@meshy @KevinEtienne Review?

File: setup.cfg (hunk 3 above)
Comment (meshy) [inline setup.cfg:5]: Oops wrong file! `:P`

Comment (LilyFirefly) [inline setup.cfg:5]: I worked that out after a bit of confusion. :wink:

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+coverage==3.7.1
+flake8==2.3.0
+flake8-import-order==0.5.1
```
Comment (meshy) [inline requirements.txt:3]: is it worth adding `wheel`?

**Comment by meshy** [comment]
Looks good to go to me

File: requirements.txt (hunk 4 above)
Comment (LilyFirefly) [inline requirements.txt:3]: Maybe to [`setup_requires`](http://pythonhosted.org/setuptools/setuptools.html#new-and-changed-setup-keywords) in `setup.py`?

**Pull request merged**